uitools/widgets/Esri/ArcGISRuntime/Toolkit/FloorFilter.cpp

```diff
@@ -0,0 +1,311 @@
+     and then sets `b` back to its initial value on destruction.
+     */
+    template <typename T>
+    struct PushValue
+    {
+      PushValue(T& b, T v) :
+        m_tracked(b),
+        m_state{std::move(m_tracked)}
+      {
+        m_tracked = std::move(v);
+      }
+
+      ~PushValue()
+      {
+        m_tracked = std::move(m_state);
+      }
+
+      T& m_tracked;
+      T m_state;
+    };
+
+    /*!
+     \internal
+     \brief Creates a `PushValue`.
+     */
+    template <typename T>
+    PushValue<T> push_value(T& t, T v)
+    {
+      return PushValue<T>(t, v);
+    }
+  }
+
+  /*!
+    \inmodule EsriArcGISRuntimeToolkit
+    \class Esri::ArcGISRuntime::Toolkit::FloorFilter
+    \brief Allows to display and filter the available floor aware layers in the current \c GeoModel.
+    The FloorFilter allows the interaction with the available floor aware layers. A user can select from a list of sites which presents
+    their facilities. Once a facility is chosen, it is possible to toggle between its levels which will show them on the \c GeoView.
+    2D maps and 3D scenes are supported.
+   */
+
+  /*!
+    \brief Constructs a new FloorFilter object with a given \a parent.
+   */
+  FloorFilter::FloorFilter(QWidget* parent) :
+    QFrame(parent),
+    m_controller(new FloorFilterController(this)),
+    m_ui(new Ui::FloorFilter)
+  {
+    m_ui->setupUi(this);
+
+    m_ui->allSites->setChecked(!m_controller->isSelectedSiteRespected());
+    m_ui->sitesView->setModel(wrapFilterModel(m_controller->sites(), m_ui->sitesFiltler, this));
+    m_ui->facilitiesView->setModel(wrapFilterModel(m_controller->facilities(), m_ui->facilitiesFilter, this));
+    m_ui->levelsView->setModel(m_controller->levels());
+
+    // Changes the contents of the facilities list.
+    connect(
+        m_ui->allSites, &QCheckBox::clicked, this, [this](bool checked)
+        {
+          m_controller->setIsSelectedSiteRespected(!checked);
+          if (checked)
+            m_ui->toolBox->setCurrentIndex(1);
+        });
+
+    // Sites setup
+    connect(
+        m_controller, &FloorFilterController::selectedSiteIdChanged, this, [this](QString /*oldId*/, QString newId)
+        {
+          // Set `m_sitesUpdatedFromController` for the duration of scope, then set back to false.
+          auto b = push_value(m_sitesUpdatedFromController, true);
+          const auto i = indexForId<FloorFilterSiteItem>(m_ui->sitesView->model(), newId);
+          m_ui->sitesView->selectionModel()->setCurrentIndex(i, QItemSelectionModel::SelectCurrent);
+        });
+
+    connect(
+        m_ui->sitesView, &QListView::doubleClicked, this, [this](QModelIndex index)
+        {
+          if (index.isValid())
+            m_ui->toolBox->setCurrentIndex(1);
+        });
+
+    connect(
+        m_ui->sitesView->selectionModel(), &QItemSelectionModel::currentChanged, this, [this](QModelIndex index, QModelIndex previous)
+        {
+          if (index == previous)
```

Comment:
💭  I noticed that the model is hooked up to the controller and when a facility is selected by panning around, this slot is being run. The operations set in this function have been already run in the controller, so it is performing twice. I would recommend to early return in case the sites were updated from the controller
```suggestion
          if (index == previous || m_facilitiesUpdatedFromController)
```

Reply:
Good find!